src/main/scala/li/cil/oc/server/component/Trade.scala

```diff
@@ -0,0 +1,256 @@
+package li.cil.oc.server.component
+
+import java.util.UUID
+
+import li.cil.oc.Settings
+import li.cil.oc.api.machine._
+import li.cil.oc.api.prefab.AbstractValue
+import li.cil.oc.common.EventHandler
+import li.cil.oc.util.InventoryUtils
+import net.minecraft.entity.passive.EntityVillager
+import net.minecraft.item.ItemStack
+import net.minecraft.nbt.NBTTagCompound
+import net.minecraft.village.MerchantRecipe
+import net.minecraftforge.common.DimensionManager
+import net.minecraftforge.common.util.ForgeDirection
+import ref.WeakReference
+import li.cil.oc.api.network.EnvironmentHost
+import scala.collection.JavaConverters._
+
+class Trade(info: TradeInfo) extends AbstractValue {
+
+  def this() = this(new TradeInfo())
+  def this(upgr: UpgradeTrading, villager: EntityVillager, recipeID: Int) = {
+    this(new TradeInfo(upgr.host, villager, recipeID))
+  }
+
+
+  override def load(nbt: NBTTagCompound) = {
+    EventHandler.scheduleServer(() => {
+      //Tell the info to load from NBT, behind EventHandler because when load is called we can't access the world yet
+      //and we need to access it to get the Robot/Drone TileEntity/Entity
+      info.load(nbt)
+    })
+  }
+
+  override def save(nbt: NBTTagCompound) = {
+    //Tell info to save to nbt
+    info.save(nbt)
+  }
+
+  def inventory = info.inventory
+
+  @Callback(doc="function():table, table -- returns the items the villager wants for this trade")
+  def getInput(context: Context, arguments: Arguments): Array[AnyRef] = {
+    Array(info.recipe.getItemToBuy.copy(),
+      info.recipe.hasSecondItemToBuy match {
+        case true => info.recipe.getSecondItemToBuy.copy()
+        case false => null
+      })
+  }
+
+  @Callback(doc = "function():table -- returns the item the villager offers for this trade")
+  def getOutput(context: Context, arguments: Arguments): Array[AnyRef] = {
+    Array(info.recipe.getItemToSell.copy())
+  }
+
+  @Callback(doc="function():boolean -- returns whether the villager currently wants to trade this")
+  def isEnabled(context: Context, arguments: Arguments): Array[AnyRef] = {
+    Array(info.villager.exists((villager: EntityVillager) => // Make sure villager is neither dead/gone nor the recipe
+      !info.recipe.isRecipeDisabled                          // has been disabled
+    ).asInstanceOf[AnyRef])
+  }
+
+  val maxRange = Settings.get.tradingRange
+  def inRange = info.villager.isDefined && distance.exists((distance: Double) => distance < maxRange)
+  def distance = info.villager match {
+    case Some(villager: EntityVillager) =>
+      info.host match {
+        case Some(h: EnvironmentHost) => Some(Math.sqrt(Math.pow(villager.posX - h.xPosition, 2) + Math.pow(villager.posY - h.yPosition, 2) + Math.pow(villager.posZ - h.zPosition, 2)))
```

Comment:
Minor nitpick (which I tend to forget myself), while in this case this is perfectly safe, I *try* to always use `Option(x)` for creating optionals, because that's always guaranteed to be safe, even if `x` is `null` - for `Some` that'd create an optional with a `null` inside, which you generally don't want.